app/models/user.rb

```diff
@@ -25,80 +29,165 @@ def to_s
     webauth.attributes['DISPLAYNAME'] || webauth.login
   end
 
-  @role_cache = {}
-
   # Queries Solr for a given record, returning synthesized role strings that are defined. Searches:
   # (1) for User by sunet id
   # (2) groups actually contains the sunetid, so it is just looking at different solr fields
   # NOTE: includes legacy roles that have to be translated
-  # @param [String] DRUID, fully qualified
-  # @return [Array[String]] list of roles
+  # @param [String] DRUID
+  # @return [Array<String>] set of roles permitted by DRUID
   def roles(pid)
+    druid = DruidTools::Druid.new(pid).druid
     @role_cache ||= {}
-    return @role_cache[pid] if @role_cache[pid]
-
-    resp = Dor::SearchService.query('id:"' + pid + '"')['response']['docs'].first || {}
-    toret = []
-    my_groups = groups
-    %w(apo_role_group_manager_ssim apo_role_person_manager_ssim).each do |key|
-      toret << 'dor-apo-manager' if resp[key] && (resp[key] & my_groups).length > 0
-    end
-
+    return @role_cache[druid] if @role_cache[druid]
+    # Try to retrieve a Solr doc for the DRUID
+    obj_doc = Dor::SearchService.query('id:"' + druid + '"')['response']['docs'].first || {}
+    return [] if obj_doc.empty?
+    druid_roles = Set.new
+    # Determine whether user has 'dor-apo-manager' role
+    apo_roles = %w(apo_role_group_manager_ssim apo_role_person_manager_ssim)
+    druid_roles << 'dor-apo-manager' if apo_roles.any? {|r| role_allowed?(obj_doc, r)}
+    # Check additional known roles
     known_roles.each do |role|
-      ["apo_role_#{role}_ssim", "apo_role_person_#{role}_ssim"].each do |key|
-        toret << role if resp[key] && (resp[key] & my_groups).length > 0
-      end
+      apo_roles = ["apo_role_#{role}_ssim", "apo_role_person_#{role}_ssim"]
+      druid_roles << role if apo_roles.any? {|r| role_allowed?(obj_doc, r)}
     end
-    @role_cache[pid] = toret # store this for now, there may be several role related calls
-    toret
+    # store and return an array of roles (Set.sort is an Array)
+    @role_cache[druid] = druid_roles.sort
   end
 
-  # @return [Array<String>] list of apos the user is allowed to view
+  # @return [Array<String>] list of roles the user can adopt
   def known_roles
-    ['dor-administrator', 'sdr-administrator', 'dor-viewer', 'sdr-viewer', 'dor-apo-creator', 'dor-apo-manager', 'dor-apo-depositor', 'dor-apo-reviewer', 'dor-apo-metadata', 'dor-apo-viewer']
+    @known_roles ||= %w(
+      dor-administrator
+      dor-apo-creator
+      dor-apo-depositor
+      dor-apo-manager
+      dor-apo-metadata
+      dor-apo-reviewer
+      dor-apo-viewer
+      dor-viewer
+      sdr-administrator
+      sdr-viewer
+    )
+  end
+
+  # Validate a user belongs to a workgroup allowed to access a DOR object
+  # @param dor_doc [Hash] A Solr document for a DOR object
+  # @param roles [Array<String>] Roles to be validated for the DOR object
+  def role_allowed?(solr_doc, role)
```

Comment:
same sort of thing here, seems like `role_field_name` or similar would be clearer.  also, the comment about `roles` is slightly off on both name and expected type, since what's expected is a single string for use as a key (right?).